### PR TITLE
[feat]: [PIPE-18845]: Parsing the webhook payload to get the modified, added and removed files list

### PR DESCRIPTION
**Pull request by adivishy1**
Code changes: 
For github and gitlab, added code to parse the webook payload for the push hook based events to get the list of modified, added and removed files. Currently we get the list of modified, added and removed files only from the github and gitlab [Bitbukcet doesnt send this information in the payload]

Testing: The code changes involved dont have an external api call, its just the parsing logic change, have tested out using the existing UTs and added new UTs as well.

Reviewers:
Mohit and Raghav

### CODE REVIEW

**Comment by bradrydzewski** [comment]
For other webhooks, we store added / modified / deleted / renamed data in the `scm.Change` structure.  Can we re-use this data structure for push webhooks as well?
https://github.com/drone/go-scm/blob/0a0caf128644b914736b718a2d63f911d6df352d/scm/pr.go#L54

Also instead of `interfaceSliceToStringSlice` I would recommend we strongly type these fields

```diff
-			Modified []interface{} `json:"modified"`
-			Removed  []interface{} `json:"removed"`
+			Modified []string `json:"modified"`
+			Removed  []string `json:"removed"`
```

**Comment by adivishy1** [comment]
> For other webhooks, we store added / modified / deleted / renamed data in the `scm.Change` structure. Can we re-use this data structure for push webhooks as well? While GitHub and GitLab only provide file names, some other vendors (like Azure, Stash) provide additional metadata that we can capture in the `scm.Change` struct
> 
> https://github.com/drone/go-scm/blob/0a0caf128644b914736b718a2d63f911d6df352d/scm/pr.go#L54-L61
> 
> Also instead of `interfaceSliceToStringSlice` I would recommend we strongly type these fields
> 
> ```diff
> -			Modified []interface{} `json:"modified"`
> -			Removed  []interface{} `json:"removed"`
> +			Modified []string `json:"modified"`
> +			Removed  []string `json:"removed"`
> ```

Question 1:
Also instead of interfaceSliceToStringSlice I would recommend we strongly type these fields
Wouldnt this be a backward incompatible change?

Question 2:
Since we get the list of modified/added/removed files in the commits part of the payload, wouldnt it be better if we return in the commits instead of the scm.Change?

Question 3:
For Azure/Stash, we are making a git call to get the list of modified/added/removed files, didnt get you how we can get this data from the metadata

**Comment by adivishy1** [comment]
Hi @bradrydzewski,
The customers who are using github as their git providers were facing rate limit issues. For these customers we have onboarded them to the BiDirectional Sync feature and inspite of this they are running into rate limits. On debugging the issue, we saw during the webhook processing, to get the list of modified/added/removed files, we are making a list commit call to Github inspite of having this details in the payload. We want to get ride of list commit call and use the payload data for getting this file details. As majority of the customers are using Github, we want to first implement this change for Github.

This is a high priority task which we want to close soon so that we can help the customer with the rate limit issue, can you pls help us with the reviews? Thanks in advance

**Comment by bradrydzewski** [comment]
> Question 1:
> Also instead of interfaceSliceToStringSlice I would recommend we strongly type these fields
> Wouldnt this be a backward incompatible change?

No, this would not be a backward incompatible because these structs are private and cannot be used outside of the package. You will note that the struct name is lowercase which means it is private ... only uppercase structs are public

```
pushHook struct {
```

**Comment by adivishy1** [comment]
> > Question 1:
> > Also instead of interfaceSliceToStringSlice I would recommend we strongly type these fields
> > Wouldnt this be a backward incompatible change?
> 
> No, this would not be a backward incompatible because these structs are private and cannot be used outside of the package. You will note that the struct name is lowercase which means it is private ... only uppercase structs are public
> 
> ```
> pushHook struct {
> ```

Got it, thanks for the explanation, have changed the type to list of string

**Comment by adivishy1** [comment]
Hi @bradrydzewski,
We are parking this optimisation for now as the this will not be giving us much benefit with the rate limits and also its limited only to GitHub and Gitlab, thanks for your inputs and support during the pr review.

File: scm/pr.go [hunk 1]
```diff
@@ -56,6 +56,7 @@ type (
 		Added        bool
 		Renamed      bool
 		Deleted      bool
+		Modified     bool
```
[the remark this answers is not in the document]
Comment (adivishy1) [reply scm/pr.go:59]: does this mean that, for modified files, we will be just adding the file path and the boolean fields, added, renamed and deleted will be false?

**Comment by bradrydzewski** [comment]
> does this mean that, for modified files, we will be just adding the file path and the boolean fields, added, renamed and deleted will be false?

Correct.

Please note that boolean values are false by default and strings are empty by default.  This means you do not need to explicitly set these values and you can modify the code as follows:

```diff
		modified = append(modified, scm.Change{
			Path:         file,
-			Added:        false,
-			Renamed:      false,
-			Deleted:      false,
			Modified:     true,
-			PrevFilePath: "",
		})
```

_EDIT I see the pull request has been parked for now so the feedback can be ignored_

**Pull request closed without merging**